### DISCUSSION TITLE
Use `collectAsStateWithLifecycle` instead of `collectAsState` for collecting flows in a lifecycle-aware manner

### DISCUSSION THREAD

**Discussion by lydiasama** [nimblehq/android-templates#398] [RFCs]
## Issue

In Jetpack Compose, we must convert any Observable type from the ViewModel to a `State<T>` so that Jetpack Compose can automatically recompose when the state changes.
And with the `Kotlin Flow`, we are using the `collectAsState` function in our Android Template to collect flows from the ViewModel in the Composable function. However, **that function is for platform-agnostic code, not for Android-only and it's not in a lifecycle-aware manner.** 

## Solution

Since the [androidx.lifecycle:lifecycle](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.6.0-alpha01) library released the new [collectAsStateWithLifecycle()](https://developer.android.com/reference/kotlin/androidx/lifecycle/compose/package-summary#extension-functions) collects values from a `Flow` in a lifecycle-aware manner, allowing your app to save unneeded app resources. It represents the latest emitted value via Compose [State](https://developer.android.com/reference/kotlin/androidx/compose/runtime/State). Use this API as the recommended way to collect flows on Android apps.

To use `collectAsStateWithLifecycle()`, we need to add one more dependency:
```kotlin
dependencies {
      ...
      implementation("androidx.lifecycle:lifecycle-runtime-compose:2.6.0-alpha03")
}
```

Whereas if we continue using the `collectAsState()`, no dependency is needed. It's included in the `compose-runtime` lib already.

The example usage is:
```kotlin
val viewState by viewModel.state.collectAsStateWithLifecycle()
```

You can check out the example from google's repository [here](https://github.com/android/compose-samples/blob/main/JetNews/app/src/main/java/com/example/jetnews/ui/interests/InterestsScreen.kt#L181)

Please take a look at the details [here](https://developer.android.com/jetpack/compose/state#use-other-types-of-state-in-jetpack-compose).
More refs from Medium: [Consuming flows safely in Jetpack Compose](https://medium.com/androiddevelopers/consuming-flows-safely-in-jetpack-compose-cde014d0d5a3)

So it comes up with this RFC to change our Android Template use of `collectAsStateWithLifecycle` instead of `collectAsState`.

## Who Benefits?

Android developers

## What's Next?

Create a PR for updating the `collectAsState` to `collectAsStateWithLifecycle`

## Extras 😄 

Why we should use collectAsStateWithLifecycle instead of collectAsState to collect flow in Android? from ChatGpt:

![Screenshot 2566-02-09 at 12 09 45 PM](https://user-images.githubusercontent.com/12026942/217723520-a3067efb-a5a7-4fcd-9fb0-c78fe8fc3b6f.png)

**Comment by kaungkhantsoe** [comment]
I found [this question](https://stackoverflow.com/questions/74259605/is-collectasstatewithlifecycle-only-applicable-to-cold-flow-and-not-helpful-for) about `collectAsState` and `collectAsStateWithLifecycle`. Can you provide an example of using `collectAsStateWithLifecycle` could be better? @lydiasama

**Reply by lydiasama** [reply]
We can use exactly what we use the `collectAsState` like this:

```kotlin
val viewState by viewModel.state.collectAsStateWithLifecycle()
```

You can see the example from google's jetpack compose repository [here](https://github.com/android/compose-samples/blob/main/JetNews/app/src/main/java/com/example/jetnews/ui/interests/InterestsScreen.kt#L181).

**Reply by kaungkhantsoe** [reply]
@lydiasama I mean will the `states` keep emitting new values even after the screen is out of the lifecycle like 
- Pausing the lifecycle by clicking on home button
- Navigating away from the screen to other screen like `Detail Screen`

I just tested both `collectAsState` and `collectAsStateWithLifecycle` using those 2 methods and `states` stopped emitting new values once I am away from the screen. May be because we are inside `viewModelScope` while emitting new values, I don't see a problem with `collectAsState`. Am I missing something? 🙇

**Reply by lydiasama** [reply]
@kaungkhantsoe I found the answer for you here: https://www.reddit.com/r/androiddev/comments/wm5s7u/comment/ikajlff/?utm_source=share&utm_medium=web2x&context=3

> viewModelScope does not get cancelled right away when you background your app though. So it can still be emitting events when your users are using other apps until Android decides to terminate your app in the background.
You can test it yourself by backgrounding the app and putting a breakpoint in the collectAsState implementation if (context == EmptyCoroutineContext) { condition. You'll see that even when your app is backgrounded that breakpoint will trigger. Using the collectAsStateWithLifecycle will make sure those flows aren't collected while your app is backgrounded.

and here: https://www.droidcon.com/2021/11/05/real-time-lifecycle-aware-updates-in-jetpack-compose/

> Lack of proper lifecycle-awareness
>
> We are subscribing, unsubscribing, everything works just fine. But there is a subtle issue. Remember the scope we’ve used to launch the flow collection in? — viewModelScope. And that’s the problem.
>
> Whenever we put the app in background by switching to a different app, or switching the screen off, price updates are still coming. UI won’t be recomposed in this scenario, but producers are still active and emit new values even when we can’t show them to the user.

**Reply by kaungkhantsoe** [reply]
@lydiasama I just debug test with a break point at `collectAsState` and it is also behaving the same as `collectAsStateWithLifecycle`. I did the following actions

1. Add a break point in a `Composable` where we collect the state with `collectAsState`
2. Debug run the app
3. Press home to put the app to background as soon as the app starts

I did the same process with `collectAsStateWithLifecycle` and it is the same. Am I missing something here? 💭 
Because it doesn't go to the break point unless I resume the app in both cases.

**Reply by lydiasama** [reply]
@kaungkhantsoe Let's see how ChatGpt answers this question. 😂 

![Screenshot 2566-02-09 at 12 09 45 PM](https://user-images.githubusercontent.com/12026942/217723678-688afaad-efdc-4500-9993-11398daabb46.png)

**Reply by lydiasama** [reply]
@kaungkhantsoe But we cannot use the breakpoint to debug our app when it's in the background state, right? Many resources confirm that it still emits while the app is in the background. So let's use the latest `collectAsStateWithLifecycle` as Google suggested to avoid the edge cases we could not test ourselves. What do you think?

**Reply by kaungkhantsoe** [reply]
@lydiasama I do support `collectAsStateWithLifecycle`. I just want to find how we can know that it is better than `collectAsState`. 😬 Let's do it. 🚀

**Reply by doannimble** [reply]
@kaungkhantsoe You can try setting the delay (10 or more seconds before emitting the result). In the meantime, close/terminate the current screen (by back press, for example) 🤔

**Reply by kaungkhantsoe** [reply]
Yes, I did @doannimble. Still it didn't go to the break point. I delayed 10 seconds and even tried using `CoroutineScope` instead of `viewModelScope`. It didn't work.

**Reply by chornerman** [reply]
@kaungkhantsoe For the POC, please check out the below videos. We can see that `collectAsState` is still collecting the flow even after we minimized the app. However, `collectAsStateWIthLifeCycle` isn't. For more information, please check out [this ref](https://medium.com/tech-takeaways/how-to-safely-collect-flows-lifecycle-aware-in-jetpack-compose-a-new-approach-ed20ead25be9)

**Collect as state**
https://user-images.githubusercontent.com/13492460/218009617-725f57b5-d602-4e97-9503-eb3cfcdfac34.mov

**Collect as state with life cycle**
https://user-images.githubusercontent.com/13492460/218009788-3a4400e2-0777-4699-9044-a19c12d8a384.mov

**Comment by doannimble** [comment]
Voted +1 👍

**Comment by thiennguyen0196** [comment]
Voted @lydiasama 👏

**Comment by luongvo** [comment]
Voted @lydiasama. Actually, according to your last raise in the channel, I applied it in my KMM's IC and it seems to work well https://github.com/luongvo/kmm-survey/blob/2ac554a171ad92a451e7450178770860c57d69de/android/src/main/java/vn/luongvo/kmm/survey/android/ui/screens/login/LoginScreen.kt#L56-L58 💪 .

**Comment by chornerman** [comment]
Voted 🚀

**Comment by Wadeewee** [comment]
Voted 🚀

**Comment by lydiasama** [comment]
@Tuubz @ryan-conway Shall we turn this into an issue? You can assign me to this task. 🙏🏻

**Comment by lydiasama** [comment]
@ryan-conway Since we got this discussion conclusion, I think we can close it. 
Ref: https://github.com/nimblehq/compass/discussions/1034
